designs/data-lake.md

```diff
@@ -0,0 +1,121 @@
+
+
+## Proposal
+
+To make it as easy as possible for data engineers to get started in developing pipelines, and to provide an existing storage structure to enforce industry best practices out-of-the-box, I propose the introduction of a datalake component to the FlowEHR platform.
+
+### What is a Data Lake?
+
+A data lake is a massively scalable and secure data storage for high-performance analytics workloads. Due to the volume of data that modern organisations need to deal with, data lakes have become a defacto choice for storing, analysing and consuming data at scale.
+
+In Azure terms, the Azure Data Lake Storage Gen2 service is a hierarchical file system built on top of Blob Storage, which provides the following benefits:
+
+- **Massive scale**: Store and process trillions of files.
+- **Cost-effective**: Pay only for what you use.
+- **Optimized performance**: High throughput and low latency.
+
+As the underlying storage is file-based, the data lake can store high quantities of multi-modal data in one place: structured data (tables - CSV, Parquet etc.), unstructured data, images, videos, etc.
+
+Integrating this with FlowEHR, alongside a number of best-practices with managing a data lake, will enable adopters of the platform to get started with data engineering in a scalable and robust way.
+
+One advantage of using Data Lake and Spark is ability to use Delta tables. They can easily be [created from Databricks Runtime](https://learn.microsoft.com/en-us/azure/databricks/delta/tutorial#--create-a-table) as well as [Synapse](https://learn.microsoft.com/en-us/azure/synapse-analytics/spark/apache-spark-delta-lake-overview?pivots=programming-language-python#create-a-table) using a statement `df.write.saveAsTable("myTableName")`. Delta Tables have versioning enabled by default which allows to track all updates to the tables. They can also [track row-level changes](https://learn.microsoft.com/en-us/azure/databricks/delta/delta-change-data-feed) which can be read downstream for incremental processing. They have many other useful features such as [automatic schema validation](https://learn.microsoft.com/en-us/azure/databricks/delta/schema-validation) and are a widespread storage format for Big Data processing.
+
+### Data Lake best practices
+
+To avoid "data swamps", where a data lake becomes a dumping ground for data without any structure or governance, there are a number of [best practices](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/scenarios/cloud-scale-analytics/best-practices/data-lake-overview#data-lake-planning) to consider, one of which is zones.
+
+Zones guarantees atomicity, consistency, isolation, and durability as data passes through multiple layers of validations and transformations before being stored in a layout optimized for consumption. Each zone is essentially its own filesystem with a heirarchical folder structure. In a medallion structure, as alluded to before, a simple implementation of zones could be:
```

Comment:
My worry is that a medallion structure sounds monolithic. That is there's just one team or group that decides what's bronze, silver or gold. Is this true? I think for the organisation to make progress, these decisions need to be delegated. That is there should be a strong incentive to load all data into one place. I assume this is 'bronze'. There should be almost no barrier to entry here.
The silver/gold build should be done by teams of users. Some of them will have official roles in the organisation, some will be temporary researchers. Silver/gold should be inspectable by _all_. Then there would be some versions of silver/gold that are owned and run by the organisation. The badge of approval is defined by who 'owns' these transformations (i.e. internal NHS business teams). But because other silver/gold work is searchable and inspectable, then these teams might choose to promote and badge other individual teams' contributions.

Reply:
The Gold / Silver / Bronze zones are just places to store data - and give an indication as to the quality / readiness of that data. Having a single data lake or single zones doesn't mean that one group 'own' all the data and in those zones, or are in charge or moving it between zones. Individual groups would still build pipelines to ingest / prep / enrich their data source - but the Bronze/Silver/Gold concept would give them the structure to work to.

I think we should maybe discuss the idea of all data in Gold being inspectable by all - outside of this PR perhaps? Be interested to know if you mean the real data, or the schemas (such as a data catalogue).